flows-advanced/etl/elt-with-snowflake.py

```diff
@@ -0,0 +1,335 @@
+    retry_delay_seconds=300,
+)
+def extract_data_from_api(start_time: dt, batchSystemCode_list: List[int]) -> DataFrame:
+
+    """
+    Given a list of depts and a start time, will retrieve needed data from API.
+
+    Examples:
+    - If there is no data in the table, and the time is
+    2022-08-25T10:30:01.000000Z, the list of startDts will look like so:
+    ["2022-07-01T00:00:01.000000Z", "2022-07-02T00:00:01.000000Z",
+     "2022-07-03T00:00:01.000000Z", ..., "2022-09-12T00:00:01.000000Z"]
+     That means we will load data from these dates, UP TO the current time.
+     We can only request one date and one dept at a time from ABC.
+
+    - If there is data in the table, and the time is
+    2022-08-25T10:30:01.000000Z, we would look at the latest ETL_LOADED_ON
+    value in the Snowflake table to see the last time data was loaded. Let's
+    say MAX(ETL_LOADED_ON) = 2022-08-23T10:30:01.000000Z, we'll add one
+    second to that, and the list of startDts will look like so:
+    ["2022-08-23T10:30:02.000000Z", "2022-08-24T00:00:01.000000Z",
+    "2022-08-25T00:00:01.000000Z"]
+
+    This enables us to load the remaining data from 2022-08-23 (but no
+    overlap), all data from 2022-08-24, and data up until now on 2022-08-25.
+
+    Args:
+        - start_time (dt): The time at which to start loading data
+        - batchSystemCode_list (list): Dept IDs, each referenced as a
+        batchSystemCode.
+
+    Returns:
+        - df (DataFrame): retrieved ABC API data.
+    """
+
+    logger = get_run_logger()
+
+    ABC_CREDENTIALS = json.loads(Secret.load("abc-credentials").get())
+
+    endpoint = ABC_CREDENTIALS["endpoint"]
+    headers = {
+        "api-user": ABC_CREDENTIALS["api-user"],
+        "api-access-code": ABC_CREDENTIALS["api-access-code"],
+    }
+
+    start_time_list_formatted = []  # list of dates on which to get data
+
+    # create window of time where we are missing data in the table that
+    # we need to retrieve from ABC API
+    todays_date = date.today()
+
+    # start_time refers to the last date(time) loaded into the Snowflake ABC_TABLE
+    next_day = (start_time + timedelta(days=1)).date()
+
+    while next_day <= todays_date:
+
+        # append it to list
+        start_time_list_formatted.append(next_day)
+        # increment day
+        next_day = next_day + timedelta(days=1)
+
+    # format list of startDt requests for ABC - must be in this format
+    start_time_list_formatted = [
+        date.strftime("%Y-%m-%d") + "T00:00:01.000000Z"
+        for date in start_time_list_formatted
+    ]
+
+    # increment startDt by 1 second
+    new_start_time = dt.strftime(
+        start_time + timedelta(seconds=1), "%Y-%m-%dT%H:%M:%S.%fZ"
+    )
+    # and add to the list to retrieve data since that time (for that day)
+    start_time_list_formatted.append(new_start_time)
+
+    logger.info(f"start_time list: {start_time_list_formatted}")
+
+    time_of_abc_request = asyncio.run(
+        _get_data_from_all_depts(
```

Comment:
Ideally one would be able to pull data SINCE a date, not be forced to make requests on a date-by-date basis. Although, I will say I have encountered that a few times. I think I could simplify that section, yeah, when I get a chance.